### DISCUSSION TITLE
can multiple openmp threads be used to simultaneously  read /write different parts of the same netcdf 4 file?

### DISCUSSION THREAD

**Discussion by TingLei-daprediction** [Unidata/netcdf-fortran#392] [Q&A] [answered]
Dear Mr. or Mrs.,
I had done some online searching and asked around for some experts and still failed to get a conclusion. 
So, your clarification is appreciated on my question: can multiple openmp threads be used to simultaneously  read /write different parts of the same netcdf 4 file? 
Thank you

**Comment by DennisHeimbigner** [comment, marked as answer]
This is not safe even if only reading data.
The reason is that there there are caches
that get modified during reading. Hence
there is the possibility of race conditions.

**Comment by TingLei-daprediction** [comment]
@DennisHeimbigner  Got it.   Thank you so much !

**Reply by TingLei-daprediction** [reply]
@DennisHeimbigner , Please allow me ask another following up question when I know nothing about the caching in netcdf IO.   For multiple thread (for example 5 threads) , If I use 5 soft symbol links pointed to the same physical file, and each thread is used to read different soft links (though they are actually reading the same physical file).  Will the netcdf create 5 different caches like it is reading  5 physically (really) different files (and, hence,  the different caches won't cause thread race situation) ?  Will this achieve the faked thread safety? 
Your further clarification is appreciated.

**Reply by DennisHeimbigner** [reply]
Frankly, I do not know. For historical reasons, analyzing the thread-safety of netcdf-c is difficult.
I suspect that the scenario you suggest will work, but I cannot guarantee it.

**Reply by TingLei-daprediction** [reply]
@DennisHeimbigner  Thank you!  Then, at least I can discuss with my colleagues and other experts on some following digging in this direction.   I will keep updating here if any findings. 
Really appreciate your help!
